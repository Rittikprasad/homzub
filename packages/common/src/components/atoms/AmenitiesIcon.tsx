import React from 'react';
import { StyleProp, StyleSheet, View, ViewStyle } from 'react-native';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import Icon from '@homzhub/common/src/assets/icon';
import { theme } from '@homzhub/common/src/styles/theme';
import { Divider } from '@homzhub/common/src/components/atoms/Divider';
import { ITypographyProps, Typography } from '@homzhub/common/src/components/atoms/Typography';

interface IProps {
  direction: 'row' | 'column';
  icon: string;
  iconSize?: number;
  iconColor?: string;
  label: string;
  labelStyles?: ITypographyProps;
  isLastIndex?: boolean;
  containerStyle?: StyleProp<ViewStyle>;
}

const AmenitiesIcon = (props: IProps): React.ReactElement => {
  const {
    direction,
    icon,
    iconSize = 40,
    iconColor = theme.colors.darkTint3,
    label,
    isLastIndex = false,
    containerStyle,
    labelStyles,
  } = props;

  const renderText = (): React.ReactElement => {
    const labelStyle = direction === 'column' ? styles.columnLabel : styles.rowLabel;
    return (
      <Typography
        variant={PlatformUtils.isWeb() ? 'label' : labelStyles ? labelStyles.variant : 'text'}
        size={PlatformUtils.isWeb() ? 'regular' : labelStyles ? labelStyles.size : 'small'}
        fontWeight="regular"
        style={[styles.label, labelStyle]}
        minimumFontScale={0.5}
        adjustsFontSizeToFit
      >
        {label}
      </Typography>
    );
  };

  const renderColumn = (): React.ReactElement => {
    return (
      <>
        <View style={[styles.columnContainer, containerStyle]}>
          <Icon name={icon} size={iconSize} color={iconColor} />
          {renderText()}
        </View>
        {!isLastIndex && direction === 'column' && <Divider containerStyles={styles.divider} />}
      </>
    );
  };

  const renderRow = (): React.ReactElement => {
    return (
      <View style={[styles.rowContainer, containerStyle]}>
        <Icon name={icon} size={iconSize} color={iconColor} />
        {renderText()}
      </View>
    );
  };

  return direction === 'column' ? renderColumn() : renderRow();
};

export { AmenitiesIcon };

const styles = StyleSheet.create({
  columnContainer: {
    flexDirection: 'column',
    justifyContent: 'space-around',
    alignItems: 'center',
    marginHorizontal: 10,
  },
  rowContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 4,
  },
  label: {
    color: theme.colors.darkTint3,
  },
  columnLabel: {
    alignItems: 'center',
  },
  rowLabel: {
    marginStart: 6,
  },
  divider: {
    borderWidth: 1,
    borderColor: theme.colors.background,
    height: 25,
  },
});
