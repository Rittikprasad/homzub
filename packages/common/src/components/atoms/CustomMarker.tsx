import React from 'react';
import { View, StyleSheet } from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';

interface IProps {
  selected: boolean;
}

const CustomMarker = (props: IProps): React.ReactElement => {
  const { selected } = props;

  if (!selected) {
    return <View style={styles.marker} />;
  }

  return (
    <View style={styles.selectedMarker}>
      <View style={styles.marker} />
    </View>
  );
};

const styles = StyleSheet.create({
  marker: {
    ...(theme.circleCSS(16) as object),
    backgroundColor: theme.colors.primaryColor,
  },
  selectedMarker: {
    ...(theme.circleCSS(40) as object),
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: theme.colors.markerOpacity,
  },
});

const memoizedComponent = React.memo(CustomMarker);
export { memoizedComponent as CustomMarker };
