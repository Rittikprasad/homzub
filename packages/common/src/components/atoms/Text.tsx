import React, { ReactElement } from 'react';
import { Text as RNText, StyleSheet, TextProps, StyleProp, TextStyle } from 'react-native';
import { I18nService } from '@homzhub/common/src/services/Localization/i18nextService';

type TextFieldType = 'label' | 'text' | 'title';
type TextSizeType = 'small' | 'regular' | 'large';
type FontWeightType = 'light' | 'regular' | 'semiBold' | 'bold' | 'extraBold';

interface IProps extends TextProps {
  children: string | React.ReactNode;
  type?: TextSizeType;
  textType?: FontWeightType;
  testID?: string;
  maxLength?: number;
}

export const fontFamilies = {
  english: {
    light: 'OpenSans-Light',
    regular: 'OpenSans-Regular',
    semiBold: 'OpenSans-SemiBold',
    bold: 'OpenSans-Bold',
    extraBold: 'OpenSans-ExtraBold',
  },
};

const fontLineHeights = {
  label: {
    small: 15,
    regular: 18,
    large: 22,
  },
  text: {
    small: 20,
    regular: 30,
    large: 26,
  },
  title: {
    regular: 41,
    large: 54,
  },
};

const fontSelection = {
  light: {
    fontFamily: I18nService.select<string>({
      rtl: fontFamilies.english.light,
      ltr: fontFamilies.english.light,
    }),
    ...I18nService.select<object>({
      rtl: {
        fontWeight: '300',
      },
      ltr: {
        fontWeight: '300',
      },
    }),
  },
  regular: {
    fontFamily: I18nService.select<string>({
      rtl: fontFamilies.english.regular,
      ltr: fontFamilies.english.regular,
    }),
    ...I18nService.select<object>({
      rtl: {
        fontWeight: '400',
      },
      ltr: {
        fontWeight: '400',
      },
    }),
  },
  semiBold: {
    fontFamily: I18nService.select<string>({
      rtl: fontFamilies.english.semiBold,
      ltr: fontFamilies.english.semiBold,
    }),
    ...I18nService.select<object>({
      rtl: {
        fontWeight: '600',
      },
      ltr: {
        fontWeight: '600',
      },
    }),
  },
  bold: {
    fontFamily: I18nService.select<string>({
      rtl: fontFamilies.english.bold,
      ltr: fontFamilies.english.bold,
    }),
    ...I18nService.select<object>({
      rtl: {
        fontWeight: '700',
      },
      ltr: {
        fontWeight: '700',
      },
    }),
  },
  extraBold: {
    fontFamily: I18nService.select<string>({
      rtl: fontFamilies.english.extraBold,
      ltr: fontFamilies.english.extraBold,
    }),
    ...I18nService.select<object>({
      rtl: {
        fontWeight: '800',
      },
      ltr: {
        fontWeight: '800',
      },
    }),
  },
};

const Label = ({
  type = 'regular',
  style,
  children,
  textType,
  maxLength,
  testID,
  ...props
}: IProps): ReactElement<RNText> => {
  let defaultStyle: object = {};
  const fontStyle: StyleProp<TextStyle> = fontSelection[textType ?? 'regular'];
  switch (type) {
    case 'large':
      defaultStyle = styles.labelLarge;
      break;
    case 'regular':
      defaultStyle = styles.labelRegular;
      break;
    case 'small':
    default:
      defaultStyle = styles.labelSmall;
      break;
  }
  const slicedText =
    maxLength && children && children.toString().length > maxLength
      ? `${children.toString().substring(0, maxLength)}...`
      : children;

  return (
    <RNText style={[defaultStyle, fontStyle, style]} {...props} testID={testID}>
      {slicedText || children}
    </RNText>
  );
};

const Text = ({
  type = 'regular',
  style,
  children,
  textType,
  testID,
  maxLength,
  ...props
}: IProps): ReactElement<RNText> => {
  let defaultStyle: object = {};
  const fontStyle: StyleProp<TextStyle> = fontSelection[textType ?? 'regular'];
  switch (type) {
    case 'regular':
      defaultStyle = styles.textRegular;
      break;
    case 'large':
      defaultStyle = styles.textLarge;
      break;
    case 'small':
    default:
      defaultStyle = styles.textSmall;
      break;
  }
  const slicedText =
    maxLength && children && children.toString().length > maxLength
      ? `${children.toString().substring(0, maxLength)}...`
      : children;

  return (
    <RNText style={[defaultStyle, fontStyle, style]} {...props} testID={testID}>
      {slicedText || children}
    </RNText>
  );
};

const Title = ({
  type = 'regular',
  style,
  children,
  textType,
  testID,
  maxLength,
  ...props
}: IProps): ReactElement<RNText> => {
  let defaultStyle: object = {};
  const fontStyle: StyleProp<TextStyle> = fontSelection[textType ?? 'regular'];
  switch (type) {
    case 'regular':
      defaultStyle = styles.titleRegular;
      break;
    case 'large':
      defaultStyle = styles.titleLarge;
      break;
    case 'small':
    default:
      defaultStyle = styles.titleSmall;
      break;
  }
  const slicedText =
    maxLength && children && children.toString().length > maxLength
      ? `${children.toString().substring(0, maxLength)}...`
      : children;

  return (
    <RNText style={[defaultStyle, fontStyle, style]} {...props} testID={testID}>
      {slicedText || children}
    </RNText>
  );
};

export { Label, Text, Title, fontLineHeights, TextFieldType, TextSizeType, FontWeightType };

const styles = StyleSheet.create({
  labelSmall: {
    fontSize: I18nService.select<number>({
      rtl: 10,
      ltr: 10,
    }),
    lineHeight: fontLineHeights.label.small,
    textAlign: 'left',
  },
  labelRegular: {
    fontSize: I18nService.select<number>({
      rtl: 12,
      ltr: 12,
    }),
    lineHeight: fontLineHeights.label.regular,
    textAlign: 'left',
  },
  labelLarge: {
    fontSize: I18nService.select<number>({
      rtl: 14,
      ltr: 14,
    }),
    lineHeight: fontLineHeights.label.large,
    textAlign: 'left',
  },
  textSmall: {
    fontSize: I18nService.select<number>({
      rtl: 16,
      ltr: 16,
    }),
    lineHeight: fontLineHeights.text.small,
    textAlign: 'left',
  },
  textRegular: {
    fontSize: I18nService.select<number>({
      rtl: 20,
      ltr: 20,
    }),
    lineHeight: fontLineHeights.text.regular,
    textAlign: 'left',
  },
  textLarge: {
    fontSize: I18nService.select<number>({
      rtl: 24,
      ltr: 24,
    }),
    lineHeight: fontLineHeights.text.large,
    textAlign: 'left',
  },
  titleSmall: {
    fontSize: I18nService.select<number>({
      rtl: 30,
      ltr: 30,
    }),
    lineHeight: fontLineHeights.title.regular,
    textAlign: 'left',
  },
  titleRegular: {
    fontSize: I18nService.select<number>({
      rtl: 32,
      ltr: 32,
    }),
    lineHeight: fontLineHeights.title.regular,
    textAlign: 'left',
  },
  titleLarge: {
    fontSize: I18nService.select<number>({
      rtl: 36,
      ltr: 36,
    }),
    lineHeight: fontLineHeights.title.regular,
    textAlign: 'left',
  },
});
