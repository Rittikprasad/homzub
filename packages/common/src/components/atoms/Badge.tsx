import React from 'react';
import { StyleProp, StyleSheet, TextStyle, View, ViewStyle } from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';
import { FontWeightType, Label } from '@homzhub/common/src/components/atoms/Text';

interface IProps {
  title: string;
  badgeColor: string;
  badgeStyle?: StyleProp<ViewStyle>;
  textType?: FontWeightType;
  titleStyle?: StyleProp<TextStyle>;
}

export const Badge = (props: IProps): React.ReactElement => {
  const { badgeColor, badgeStyle, title, textType = 'semiBold', titleStyle = {} } = props;
  return (
    <View style={[styles.badge, badgeStyle, { backgroundColor: badgeColor }]}>
      <Label type="regular" textType={textType} style={[styles.title, titleStyle]}>
        {title}
      </Label>
    </View>
  );
};

const styles = StyleSheet.create({
  badge: {
    borderRadius: 2,
    paddingHorizontal: 6,
  },
  title: {
    color: theme.colors.white,
    paddingBottom: 0,
    alignSelf: 'center',
  },
});
