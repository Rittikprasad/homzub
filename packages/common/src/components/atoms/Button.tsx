import React from 'react';
import {
  GestureResponderEvent,
  ImageStyle,
  StyleProp,
  StyleSheet,
  TextProps,
  TextStyle,
  TouchableOpacity,
  ViewStyle,
} from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon from '@homzhub/common/src/assets/icon';
import { FontWeightType, Label, Text, TextFieldType, TextSizeType } from '@homzhub/common/src/components/atoms/Text';

export type ButtonType = 'primary' | 'secondary' | 'secondaryOutline' | 'text';

export interface IButtonProps {
  type: ButtonType;
  node?: React.ReactNode;
  children?: React.ReactNode;
  title?: string;
  onPress?: (event?: GestureResponderEvent) => void;
  disabled?: boolean;
  activeOpacity?: number;
  titleStyle?: StyleProp<TextStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  textType?: TextFieldType;
  textSize?: TextSizeType;
  fontType?: FontWeightType;
  icon?: string;
  iconSize?: number;
  iconColor?: string;
  iconStyle?: StyleProp<ImageStyle>;
  testID?: string;
  maxLength?: number;
  textStyle?: TextStyle;
  numOfLines?: number;
  ellipsizeMode?: TextProps['ellipsizeMode'];
}

export class Button extends React.PureComponent<IButtonProps> {
  public render = (): React.ReactElement => {
    const {
      onPress,
      disabled = false,
      activeOpacity = 0.5,
      title,
      icon,
      iconSize,
      iconColor,
      iconStyle,
      testID,
      node,
      children,
    } = this.props;
    return (
      <TouchableOpacity
        activeOpacity={activeOpacity}
        onPress={onPress}
        disabled={disabled}
        style={this.getContainerStyle()}
        testID={testID}
      >
        {title && this.getTextField()}
        {!!icon && <Icon name={icon} size={iconSize} color={iconColor} style={iconStyle} />}
        {node && node}
        {children && children}
      </TouchableOpacity>
    );
  };

  private getTextField = (): React.ReactElement => {
    const {
      textType,
      title,
      textSize = 'small',
      fontType = 'semiBold',
      maxLength,
      textStyle,
      numOfLines,
      ellipsizeMode = 'tail',
    } = this.props;
    let TextField = Text;

    if (textType === 'label') {
      TextField = Label;
    }

    return (
      <TextField
        type={textSize}
        textType={fontType}
        style={[this.getTextStyle(), textStyle && textStyle]}
        maxLength={maxLength}
        ellipsizeMode={ellipsizeMode}
        numberOfLines={numOfLines}
      >
        {title}
      </TextField>
    );
  };

  private getContainerStyle = (): ViewStyle => {
    const { disabled, type, containerStyle = {} } = this.props;
    let themedStyle: StyleProp<ViewStyle> = styles.primary;
    if (disabled) {
      themedStyle = styles.disabled;
    }

    if (type === 'secondary' && !disabled) {
      themedStyle = styles.secondary;
    }

    if (type === 'secondaryOutline' && !disabled) {
      themedStyle = styles.secondaryOutline;
    }

    if (type === 'text') {
      themedStyle = styles.textButton;
    }

    return StyleSheet.flatten([styles.container, themedStyle, containerStyle]);
  };

  private getTextStyle = (): TextStyle => {
    const { type, disabled = false, titleStyle = {} } = this.props;
    let themedStyle = {};
    if (type === 'secondary' && !disabled) {
      themedStyle = { color: theme.colors.active };
    }
    if (type === 'text' && !disabled) {
      themedStyle = { color: theme.colors.darkTint2 };
    }
    if (type === 'text' && disabled) {
      themedStyle = { color: theme.colors.darkTint8 };
    }

    return StyleSheet.flatten([styles.textStyle, themedStyle, titleStyle]);
  };
}

const styles = StyleSheet.create({
  container: {
    ...theme.globalStyles.center,
    borderRadius: 4,
  },
  textStyle: {
    textAlign: 'center',
    color: theme.colors.white,
    marginVertical: 12,
    marginHorizontal: 18,
  },
  disabled: {
    backgroundColor: theme.colors.disabled,
  },
  primary: {
    backgroundColor: theme.colors.primaryColor,
  },
  secondary: {
    borderWidth: 1,
    borderColor: theme.colors.active,
    backgroundColor: theme.colors.white,
  },
  secondaryOutline: {
    borderWidth: 0.5,
    borderColor: theme.colors.white,
  },
  textButton: {
    borderWidth: 0,
  },
});
