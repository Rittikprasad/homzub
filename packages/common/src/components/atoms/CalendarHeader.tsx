import React from 'react';
import { StyleSheet, View } from 'react-native';
import moment from 'moment';
import { FunctionUtils } from '@homzhub/common/src/utils/FunctionUtils';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { theme } from '@homzhub/common/src/styles/theme';
import { Divider } from '@homzhub/common/src/components/atoms/Divider';
import { Text } from '@homzhub/common/src/components/atoms/Text';

interface IProps {
  isAllowPastDate?: boolean;
  headerTitle?: string;
  headerYear?: string;
  yearTitle?: string;
  isMonthView?: boolean;
  isYearView?: boolean;
  isCurrentMonth?: boolean;
  maxDate?: string;
  minDate?: string;
  month?: number;
  year?: string;
  onBackPress?: () => void;
  onNextPress?: () => void;
  onMonthPress?: () => void;
  onYearPress?: () => void;
}

const CalendarHeader = (props: IProps): React.ReactElement => {
  const {
    isAllowPastDate,
    maxDate,
    minDate,
    onBackPress,
    onNextPress,
    onMonthPress,
    onYearPress,
    isCurrentMonth,
    headerTitle,
    headerYear,
    isMonthView = false,
    isYearView = false,
    yearTitle,
    month,
    year,
  } = props;

  // DISABLE VALIDATION
  const isNextDisable =
    maxDate &&
    (moment(maxDate).month() === moment().month() ||
      (!!month && moment(maxDate).month() === month && !!year && moment(maxDate).year() === Number(year)));
  const isPreviousDisable =
    (minDate ? moment(minDate).year() === moment().year() : !isAllowPastDate) && isCurrentMonth && !isYearView;
  // DISABLE VALIDATION

  return (
    <>
      <View style={styles.headerContainer}>
        <Icon
          name={icons.leftArrow}
          onPress={!isPreviousDisable ? onBackPress : FunctionUtils.noop}
          size={22}
          color={isPreviousDisable ? theme.colors.disabled : theme.colors.primaryColor}
        />
        {!isYearView && !isMonthView && (
          <Text type="small" textType="semiBold" onPress={onMonthPress} style={styles.headerTitle}>
            {headerTitle}
          </Text>
        )}
        {isMonthView && (
          <Text type="small" textType="semiBold" onPress={onYearPress} style={styles.headerTitle}>
            {headerYear}
          </Text>
        )}
        {isYearView && (
          <Text type="small" textType="semiBold" style={styles.headerTitle}>
            {yearTitle}
          </Text>
        )}
        <Icon
          name={icons.rightArrow}
          size={22}
          color={isNextDisable ? theme.colors.disabled : theme.colors.primaryColor}
          onPress={!isNextDisable ? onNextPress : FunctionUtils.noop}
        />
      </View>
      <Divider />
    </>
  );
};

export default CalendarHeader;

const styles = StyleSheet.create({
  headerContainer: {
    justifyContent: 'space-between',
    flexDirection: 'row',
    paddingVertical: 20,
    paddingHorizontal: 24,
  },
  headerTitle: {
    color: theme.colors.primaryColor,
  },
});
