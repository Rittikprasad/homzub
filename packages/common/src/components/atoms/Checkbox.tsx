import React from 'react';
import { StyleProp, StyleSheet, TextStyle, TouchableOpacity, ViewStyle } from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Label, TextSizeType } from '@homzhub/common/src/components/atoms/Text';

interface ICheckboxOptions {
  selected: boolean;
  label?: string;
  onToggle: () => void;
  containerStyle?: StyleProp<ViewStyle>;
  labelStyle?: StyleProp<TextStyle>;
  iconSize?: number;
  iconStyle?: object;
  iconSelectedStyle?: object;
  labelType?: TextSizeType;
  testID?: string;
  subLabel?: string;
  onPressLink?: () => void;
}

const RNCheckbox = (props: ICheckboxOptions): React.ReactElement => {
  const {
    label,
    selected,
    containerStyle = {},
    labelStyle = {},
    iconSize = 22,
    iconStyle = {},
    iconSelectedStyle = {},
    labelType = 'large',
    onToggle,
    testID,
    subLabel,
    onPressLink,
  } = props;
  return (
    <TouchableOpacity style={[styles.container, containerStyle]} onPress={onToggle} testID={testID}>
      <Icon
        name={selected ? icons.checkboxOn : icons.checkboxOff}
        size={iconSize}
        color={selected ? theme.colors.primaryColor : theme.colors.disabled}
        style={selected ? iconSelectedStyle : iconStyle}
      />
      {!!label && (
        <Label type={labelType} textType="regular" style={[styles.label, labelStyle]}>
          {label}
          {!!subLabel && (
            <Label type="regular" style={styles.subLabel} onPress={onPressLink}>
              {subLabel}
            </Label>
          )}
        </Label>
      )}
    </TouchableOpacity>
  );
};

export { RNCheckbox };

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
  },
  label: {
    marginStart: 12,
    marginEnd: 4,
    flex: 1,
  },
  subLabel: {
    color: theme.colors.primaryColor,
  },
});
