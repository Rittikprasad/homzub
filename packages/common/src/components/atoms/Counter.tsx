import React, { useEffect, useState } from 'react';
import { StyleProp, StyleSheet, TextStyle, View, ViewStyle, TextInput, Image } from 'react-native';
import { useTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { SVGUri } from '@homzhub/common/src/components/atoms/Svg';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface ITitle {
  title: string;
  titleStyle?: TextStyle;
  id?: number;
}

export interface ICounterProps {
  defaultValue: number;
  name?: ITitle;
  svgImage?: string;
  onValueChange: (count: number, id?: number) => void;
  maxCount?: number;
  minCount?: number;
  containerStyles?: StyleProp<ViewStyle>;
  testID?: string;
  disabled?: boolean;
}

export const Counter = (props: ICounterProps): React.ReactElement => {
  const { t } = useTranslation(LocaleConstants.namespacesKey.common);
  const { onValueChange, defaultValue, name, svgImage, maxCount = 10, minCount = 0, containerStyles, disabled } = props;
  const [count, setCount] = useState(defaultValue);
  const [previousCount, setPreviousCount] = useState(defaultValue);

  useEffect(() => {
    onValueChange(count, name?.id);
  }, [count]);

  const isDecDisabled = disabled || count <= minCount;
  const isIncDisabled = disabled || count >= maxCount;

  const incrementCount = (): void => {
    if (count < maxCount) {
      setCount((prev) => prev + 1);
      return;
    }
    AlertHelper.error({ message: t('maxCount', { count: maxCount }) });
  };

  const decrementCount = (): void => {
    if (count > minCount) {
      setCount((prev) => prev - 1);
      return;
    }
    AlertHelper.error({ message: t('minCount', { count: minCount }) });
  };

  const onCountChange = (value: string): void => {
    if (Number(value) > maxCount) {
      AlertHelper.error({ message: t('maxCount', { count: maxCount }) });
      setCount(Number(defaultValue));
      return;
    }
    if (Number(value) < 0) {
      AlertHelper.error({ message: t('minCount') });
      return;
    }
    setPreviousCount(Number(value));
    setCount(Number(value));
  };

  const onEndEditing = (): void => {
    if (previousCount) {
      setCount(previousCount);
      return;
    }

    if (count === 0 || count > maxCount) {
      setCount(defaultValue);
    }
  };

  const onFocus = (): void => {
    setPreviousCount(count);
    setCount(0);
  };

  return (
    <View style={[styles.rowStyle, containerStyles]}>
      <View style={styles.imageContainer}>
        {svgImage && PlatformUtils.isMobile() ? (
          <SVGUri height={24} width={24} uri={svgImage} style={styles.svgStyle} />
        ) : (
          svgImage && <Image source={{ uri: svgImage }} style={styles.imageStyle} />
        )}
        {name && (
          <Text style={[styles.textStyle, name.titleStyle]} type="small">
            {name.title}
          </Text>
        )}
      </View>
      <View style={[styles.counterContainer, styles.rowStyle]}>
        <View style={[customStyles.iconStyle(isDecDisabled), styles.iconContainer]}>
          <Icon
            name={icons.minus}
            color={isDecDisabled ? theme.colors.disabled : theme.colors.primaryColor}
            size={20}
            onPress={disabled ? undefined : decrementCount}
          />
        </View>
        <TextInput
          style={[customStyles.textInputStyle(isDecDisabled, isIncDisabled), styles.counterValue]}
          keyboardType="number-pad"
          maxLength={2}
          defaultValue={defaultValue.toString()}
          value={count.toString()}
          onChangeText={onCountChange}
          onEndEditing={onEndEditing}
          onFocus={onFocus}
        />
        <View style={[customStyles.iconStyle(isIncDisabled), styles.iconContainer]}>
          <Icon
            name={icons.plus}
            color={isIncDisabled ? theme.colors.disabled : theme.colors.primaryColor}
            size={20}
            onPress={disabled ? undefined : incrementCount}
          />
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  rowStyle: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  counterContainer: {
    borderColor: theme.colors.darkTint10,
    borderWidth: 1,
    borderRadius: 4,
    width: 104,
    padding: 4,
  },
  imageContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  iconContainer: {
    borderRadius: 2,
  },
  svgStyle: {
    marginEnd: 12,
  },
  textStyle: {
    width: PlatformUtils.isMobile() ? 100 : '100%',
  },
  counterValue: {
    textAlign: 'center',
    padding: 0,
    width: 50,
  },
  imageStyle: {
    marginRight: 12,
    width: 24,
    height: 24,
  },
});

const customStyles = {
  iconStyle: (isDecDisabled: boolean): StyleProp<ViewStyle> => ({
    backgroundColor: isDecDisabled ? theme.colors.disabledOpacity : theme.colors.activeOpacity,
  }),
  textInputStyle: (isDecDisabled: boolean, isIncDisabled: boolean): StyleProp<TextStyle> => ({
    color: isDecDisabled || isIncDisabled ? theme.colors.darkTint5 : theme.colors.primaryColor,
  }),
};
