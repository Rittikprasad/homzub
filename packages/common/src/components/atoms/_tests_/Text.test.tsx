import React from 'react';
import renderer from 'react-test-renderer';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';

const testProps = (Props: object): any => ({
  ...Props,
});
let props: any;

describe('Label Component', () => {
  describe('Snapshot tests for <Label>', () => {
    const labelTypes = [
      { type: null, name: 'Default' },
      { type: 'small', name: 'Small' },
      { type: 'regular', name: 'Regular' },
    ];

    labelTypes.forEach((labelType) => {
      it(`Type: ${labelType.name}`, () => {
        props = testProps({
          type: labelType.type,
        });
        const tree = renderer.create(<Label {...props}>Text</Label>).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});

describe('Text Component', () => {
  describe('Snapshot tests for <Text>', () => {
    const textTypes = [
      { type: null, name: 'Default' },
      { type: 'small', name: 'Small' },
      { type: 'regular', name: 'Regular' },
      { type: 'large', name: 'Large' },
    ];

    textTypes.forEach((textType) => {
      it(`Type: ${textType.name}`, () => {
        props = testProps({
          type: textType.type,
        });
        const tree = renderer.create(<Text {...props}>Demo</Text>).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
