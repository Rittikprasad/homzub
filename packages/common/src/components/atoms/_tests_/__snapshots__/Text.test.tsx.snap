// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Component Snapshot tests for <Label> Type: Default 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 10,
        "lineHeight": 15,
        "textAlign": "left",
      },
      Object {
        "fontFamily": "OpenSans-Regular",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Label Component Snapshot tests for <Label> Type: Regular 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 12,
        "lineHeight": 18,
        "textAlign": "left",
      },
      Object {
        "fontFamily": "OpenSans-Regular",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Label Component Snapshot tests for <Label> Type: Small 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 10,
        "lineHeight": 15,
        "textAlign": "left",
      },
      Object {
        "fontFamily": "OpenSans-Regular",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Text Component Snapshot tests for <Text> Type: Default 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 16,
        "lineHeight": 20,
        "textAlign": "left",
      },
      Object {
        "fontFamily": "OpenSans-Regular",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Demo
</Text>
`;

exports[`Text Component Snapshot tests for <Text> Type: Large 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 24,
        "lineHeight": 26,
        "textAlign": "left",
      },
      Object {
        "fontFamily": "OpenSans-Regular",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Demo
</Text>
`;

exports[`Text Component Snapshot tests for <Text> Type: Regular 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 20,
        "lineHeight": 30,
        "textAlign": "left",
      },
      Object {
        "fontFamily": "OpenSans-Regular",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Demo
</Text>
`;

exports[`Text Component Snapshot tests for <Text> Type: Small 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 16,
        "lineHeight": 20,
        "textAlign": "left",
      },
      Object {
        "fontFamily": "OpenSans-Regular",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Demo
</Text>
`;
